input-events/input-events-typing.html

```diff
@@ -1,82 +1,294 @@
     const expectedResult = [
         // Pressing 'a'.
         'plain-keydown-a',
         'plain-keypress-a',
-        'plain-beforeinput-a',
-        'plain-input-a',
+        'plain-beforeinput-a-null',
+        'plain-input-a-null',
         'plain-keyup-a',
         // Pressing Shift-'b'.
         'plain-keydown-B',
         'plain-keypress-B',
-        'plain-beforeinput-B',
-        'plain-input-B',
+        'plain-beforeinput-B-null',
+        'plain-input-B-null',
         'plain-keyup-B',
         // Pressing 'c'.
         'rich-keydown-c',
         'rich-keypress-c',
-        'rich-beforeinput-c',
-        'rich-input-c',
+        'rich-beforeinput-c-null',
+        'rich-input-c-null',
         'rich-keyup-c',
         // Pressing Shift-'d'.
         'rich-keydown-D',
         'rich-keypress-D',
-        'rich-beforeinput-D',
-        'rich-input-D',
+        'rich-beforeinput-D-null',
+        'rich-input-D-null',
         'rich-keyup-D',
     ];
-    let eventCounter = 0;
+    const result = [];
 
-    for (const targetId of ['plain', 'rich']) {
+    for (const target of [plain, rich]) {
         for (const eventType of ['beforeinput', 'input', 'keydown', 'keypress', 'keyup']) {
-            document.getElementById(targetId).addEventListener(eventType, t.step_func(event => {
-                if (event.key === 'Shift')
-                    return;
-                assert_equals(`${targetId}-${event.type}-${event.data || event.key}`, expectedResult[eventCounter]);
-                if (event instanceof InputEvent) {
-                    assert_equals(event.dataTransfer, null,
-                        'Plain text editing should have null |dataTransfer| field')
-                }
-                ++eventCounter;
-                if (eventCounter === expectedResult.length)
-                    t.done();
-            }));
+            target.addEventListener(eventType, event => {
+                if (event.key === 'Shift') return;
+                let eventInfo = [target.id, event.type, event.data || event.key];
+                if (event instanceof InputEvent) eventInfo.push(String(event.dataTransfer));
+                result.push(eventInfo.join('-'));
+            });
         }
     }
-    document.getElementById('plain').focus();
-    new test_driver.Actions()
-       .keyDown('a')
-       .keyUp('a')
-       .keyDown('\uE008') // Shift
-       .keyDown('B')
-       .keyUp('B')
-       .keyUp('\uE008')
-       .send()
-       .then(() => {
-           document.getElementById('rich').focus();
-           new test_driver.Actions()
-               .keyDown('c')
-               .keyUp('c')
-               .keyDown('\uE008') // Shift
-               .keyDown('D')
-               .keyUp('D')
-               .keyUp('\uE008')
-               .send();
-       });
-});</script>
+
+    plain.focus();
+    await new test_driver.Actions()
+        .keyDown('a')
+        .keyUp('a')
+        .keyDown('\uE008') // Shift
+        .keyDown('B')
```

Comment:
Maybe silly question, but could you send `.keyDown('b')` with a lowercase "b" here? Given that `shift` is pressed down, wouldn't it end up as a "B" (uppercase)?

Reply:
I would think so, yes. I don't exactly know why the author wrote it the way he did, but maybe there is no technical difference and this way he thinks he shows those reading the code how that he is expected an uppercase B? I'm just guessing. Should we change it to lowercase 'b'?

Reply:
Per [the webdriver spec](https://w3c.github.io/webdriver/#dfn-dispatch-a-keydown-action) it normalizes 'B' to 'b', so either way should be equivalent.